engine/src/main/java/org/terasology/world/block/BlockLifecycleEvent.java

```diff
@@ -32,11 +36,46 @@ public BlockLifecycleEvent(TIntList positions, BlockEntityRegistry registry) {
         this.positions = positions;
     }
 
-    public Iterable<Vector3i> getBlockPositions() {
-        return () -> new BlockPositionIterator(positions, registry);
+    @NotNull
+    @Override
+    public Iterator<Vector3ic> iterator() {
+        if (positions.size() < 3) {
+            return Collections.emptyIterator();
+        }
+        return new Iterator<Vector3ic>() {
+            private Vector3i next = new Vector3i(
+                positions.get(0),
+                positions.get(1),
+                positions.get(2));
+            private final Vector3i current = new Vector3i();
+            private int index = 3;
+            @Override
+            public boolean hasNext() {
+                return next != null;
+            }
+
+            @Override
+            public Vector3ic next() {
+                current.set(next);
+                fetchNext();
+                return current;
+            }
+
+            private void fetchNext() {
+                while (index < positions.size() - 2) {
```

Comment:
If the condition `!registry.hasPermanentBlockEntity(next)` does not hold for any position, then we will just iterate to the end and set `next = null`. If it does hold for some of the positions, then we only iterate over a subset of the positions.

So, it could be that `event.blockCount() == 6` but `for (Vector3ic pos: event)` only iterates over `4` blocks? :thinking:

Reply:
Maybe we should not inline the iterator here, but rather make it explicit that it will only iterator over `blockPositionsWithNonPermanentEntities`.